Separate affine_channel from math and optimize it

xiaomengy opened the pull request:
Summary:
Separate affine_channel from math and optimize it

i-am-not-moving-c2-to-c10

Differential Revision: D13727606

BIT-silence commented:
Benchmark:
N = 256, C = 256, H = 128, W = 128

AffineChannelOp NCHW order: 67.47 -> 38.32ms
AffineChannelOp NHWC order: 103.95-> 39.49ms

SpatialBN NCHW testing: 38.31ms, CuDNN 43.31ms
SpatialBN NCHW training: 69.06ms, CuDNN 66.58ms

houseroad commented on caffe2/utils/math/elementwise_level2.h:
```diff
@@ -0,0 +1,28 @@
+#ifndef CAFFE2_UTILS_MATH_ELEMENTWISE_H_
+#define CAFFE2_UTILS_MATH_ELEMENTWISE_H_
+
```
Can we have a more meaningful name for this header/cc file? And add some comments talking about which functions should be put in this file?

BIT-silence replied:
The AffineChannel is a special case of elementwise mul and add operation, so I think it's fine to make it in elementwise.h

houseroad commented:
The change breaks the hip build. @bddppq any suggestion?

xiaomengy committed:
Separate affine_channel from math and optimize it (#16135)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16135

Separate affine_channel from math and optimize it

i-am-not-moving-c2-to-c10

Differential Revision: D13727606

fbshipit-source-id: f7f95001075a90926d278bfd928475a4346bb3b1

houseroad reviewed (approved):
LGTM
